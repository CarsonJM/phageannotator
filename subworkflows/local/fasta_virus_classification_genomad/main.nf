//
// Classify and annotate sequences with geNomad
//

include { GENOMAD_DOWNLOAD   } from '../../../modules/nf-core/genomad/download/main'
include { GENOMAD_ENDTOEND   } from '../../../modules/nf-core/genomad/endtoend/main'
include { AWK as AWK_GENOMAD } from '../../../modules/local/awk/main'

workflow FASTA_VIRUS_CLASSIFICATION_GENOMAD {
    take:
    fasta_gz    // [ [ meta] , fasta.gz ], assemblies/genomes (mandatory)

    main:
    ch_versions = Channel.empty()

    // if genomad_db != null, skip GENOMAD_DOWNLOAD
    if ( params.genomad_db ){
        ch_genomad_db = file(params.genomad_db, checkIfExists: true)
    }
    // if genomad_db == null, run GENOMAD_DOWNLOAD
    else {
        //
        // MODULE: download geNomad database
        //
        ch_genomad_db = GENOMAD_DOWNLOAD ( ).genomad_db
        ch_versions = ch_versions.mix(GENOMAD_DOWNLOAD.out.versions.first())
    }

    //
    // MODULE: Classify/annotate viral sequences
    //
    ch_viruses_fasta_gz = GENOMAD_ENDTOEND ( fasta_gz, ch_genomad_db ).virus_fasta
    ch_versions = ch_versions.mix(GENOMAD_ENDTOEND.out.versions.first())

    //
    // MODULE: Combine geNomad summaries across samples
    //
    ch_virus_summaries_tsv = AWK_GENOMAD ( GENOMAD_ENDTOEND.out.virus_summary.map { [ [ id:'all_samples' ], it[1] ] } ).file_out
    ch_versions = ch_versions.mix(AWK_GENOMAD.out.versions.first())

    emit:
    viruses_fasta_gz = ch_viruses_fasta_gz  // [ [ meta ], fasta.gz ]   , FASTA file containing viral sequences
    versions = ch_versions                  // [ versions.yml ]

}
