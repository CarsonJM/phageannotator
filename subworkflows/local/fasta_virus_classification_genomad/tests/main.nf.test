nextflow_workflow {

    name "Test Subworkflow: FASTA_VIRUS_CLASSIFICATION_GENOMAD"
    script "subworkflows/local/fasta_virus_classification_genomad/main.nf"
    workflow "FASTA_VIRUS_CLASSIFICATION_GENOMAD"
    tag "subworkflows"
    tag "subworkflows_local"
    tag "fasta_virus_classification_genomad"
    tag "fasta_virus_classification_genomad_default"


    test("Parameters: default") {

        when {
            params {
                genomad_splits = 5
            }
            workflow {
                """
                input[0] = Channel.of(
                    [ [ id:'test' ], [ file('https://github.com/nf-core/test-datasets/raw/modules/data/genomics/prokaryotes/bacteroides_fragilis/genome/genome.fna.gz', checkIfExists: true) ] ],
                    [ [ id:'test2' ], [ file('https://github.com/nf-core/test-datasets/raw/modules/data/genomics/prokaryotes/bacteroides_fragilis/genome/genome.fna.gz', checkIfExists: true) ] ]
                )
                input[1] = null
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }
}
