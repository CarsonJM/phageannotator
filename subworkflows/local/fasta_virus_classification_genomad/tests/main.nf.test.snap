{
    "": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "genome_virus.fna.gz:md5,c2d03e27ae27e3eedd037621ee77c917"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "genome_virus_summary.tsv:md5,6899de91e735b5b930347d9d0fda8a7a"
                    ]
                ],
                "2": [
                    "versions.yml:md5,366d0d975c262aadb6a369f0bce340bf",
                    "versions.yml:md5,6b3b022e67e97057826d98ef627a82fb"
                ],
                "versions": [
                    "versions.yml:md5,366d0d975c262aadb6a369f0bce340bf",
                    "versions.yml:md5,6b3b022e67e97057826d98ef627a82fb"
                ],
                "virus_summaries_tsv": [
                    [
                        {
                            "id": "test"
                        },
                        "genome_virus_summary.tsv:md5,6899de91e735b5b930347d9d0fda8a7a"
                    ]
                ],
                "viruses_fna_gz": [
                    [
                        {
                            "id": "test"
                        },
                        "genome_virus.fna.gz:md5,c2d03e27ae27e3eedd037621ee77c917"
                    ]
                ]
            }
        ],
        "timestamp": "2023-09-26T21:30:56+0000"
    }
}